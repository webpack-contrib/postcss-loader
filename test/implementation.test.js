import postcss from "postcss";

import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from "./helpers";

describe('"implementation" option', () => {
  it("should work with implementation is string", async () => {
    const compiler = getCompiler("./css/index.js", {
      implementation: require.resolve("postcss"),
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle("style.css", stats);

    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should throw error when unresolved package", async () => {
    const compiler = getCompiler("./css/index.js", {
      implementation: "unresolved",
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats, true)).toMatchSnapshot("errors");
  });

  it("should work with a custom instance of PostCSS", async () => {
    const spy = jest.fn(postcss);
    const compiler = getCompiler("./css/index.js", {
      // Wrap the spy so it is an instanceof Function
      implementation: (...args) => spy(...args),
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle("style.css", stats);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
