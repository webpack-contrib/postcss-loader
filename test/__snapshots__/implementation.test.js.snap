// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"implementation" option should throw error when unresolved package: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
NonErrorEmittedError: (Emitted value instead of an instance of Error) Error: Cannot find module 'unresolved' from 'src/utils.js'",
]
`;

exports[`"implementation" option should throw error when unresolved package: warnings 1`] = `[]`;

exports[`"implementation" option should work with a custom instance of PostCSS: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"implementation" option should work with a custom instance of PostCSS: errors 1`] = `[]`;

exports[`"implementation" option should work with a custom instance of PostCSS: warnings 1`] = `[]`;

exports[`"implementation" option should work with implementation is string: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"implementation" option should work with implementation is string: errors 1`] = `[]`;

exports[`"implementation" option should work with implementation is string: warnings 1`] = `[]`;
